File: packages/server/src/routes/matches.ts
import { Router, type Request, type Response } from 'express'
import type { Server as SocketIOServer } from 'socket.io'
import type {
  ServerToClientEvents,
  ClientToServerEvents,
  InterServerEvents,
  SocketData,
} from '@obs-scoring/shared'
import { matchService } from '../services/matchService.js'
import { createMatchSchema, updateScoreSchema, idParamSchema } from '../lib/validation.js'

type TypedIO = SocketIOServer<ClientToServerEvents, ServerToClientEvents, InterServerEvents, SocketData>

export function createMatchRoutes(io: TypedIO): Router {
  const router = Router()

  /**
   * @swagger
   * components:
   *   schemas:
   *     Match:
   *       type: object
   *       properties:
   *         id:
   *           type: integer
   *         player1:
   *           type: string
   *         player2:
   *           type: string
   *         team1:
   *           type: string
   *           nullable: true
   *         team2:
   *           type: string
   *           nullable: true
   *         player1Score:
   *           type: integer
   *         player2Score:
   *           type: integer
   *         isCurrent:
   *           type: boolean
   *         createdAt:
   *           type: string
   *           format: date-time
   *     CreateMatch:
   *       type: object
   *       required:
   *         - player1
   *         - player2
   *       properties:
   *         player1:
   *           type: string
   *         player2:
   *           type: string
   *         team1:
   *           type: string
   *         team2:
   *           type: string
   *     UpdateScore:
   *       type: object
   *       required:
   *         - player1Score
   *         - player2Score
   *       properties:
   *         player1Score:
   *           type: integer
   *         player2Score:
   *           type: integer
   *     OBSOverlayData:
   *       type: object
   *       properties:
   *         players:
   *           type: array
   *           items:
   *             type: object
   *             properties:
   *               name:
   *                 type: string
   *               team:
   *                 type: string
   *                 nullable: true
   *               score:
   *                 type: integer
   *         matchId:
   *           type: integer
   *           nullable: true
   */

  /**
   * @swagger
   * /matches:
   *   get:
   *     summary: Get all matches
   *     tags: [Matches]
   *     responses:
   *       200:
   *         description: List of all matches
   *         content:
   *           application/json:
   *             schema:
   *               type: array
   *               items:
   *                 $ref: '#/components/schemas/Match'
   */
  router.get('/', async (_req: Request, res: Response) => {
    const matches = await matchService.findAll()
    res.json(matches)
  })

  /**
   * @swagger
   * /matches/current:
   *   get:
   *     summary: Get the current match
   *     tags: [Matches]
   *     responses:
   *       200:
   *         description: The current match
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/Match'
   *       404:
   *         description: No current match set
   */
  router.get('/current', async (_req: Request, res: Response) => {
    const match = await matchService.findCurrent()
    if (!match) {
      res.status(404).json({ error: 'No current match set' })
      return
    }
    res.json(match)
  })

  /**
   * @swagger
   * /matches/obsData:
   *   get:
   *     summary: Get OBS overlay data for current match
   *     description: Returns data formatted for OBS URL/API Source plugin
   *     tags: [OBS]
   *     responses:
   *       200:
   *         description: OBS overlay data
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/OBSOverlayData'
   */
  router.get('/obsData', async (_req: Request, res: Response) => {
    const data = await matchService.getOBSData()
    res.json(data)
  })

  /**
   * @swagger
   * /matches/{id}:
   *   get:
   *     summary: Get a match by ID
   *     tags: [Matches]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: integer
   *     responses:
   *       200:
   *         description: The match
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/Match'
   *       404:
   *         description: Match not found
   */
  router.get('/:id', async (req: Request, res: Response) => {
    const { error, value } = idParamSchema.validate({ id: parseInt(req.params.id ?? '', 10) })
    if (error) {
      res.status(400).json({ error: error.details[0]?.message })
      return
    }

    const match = await matchService.findById(value.id)
    if (!match) {
      res.status(404).json({ error: 'Match not found' })
      return
    }
    res.json(match)
  })

  /**
   * @swagger
   * /matches:
   *   post:
   *     summary: Create a new match
   *     tags: [Matches]
   *     requestBody:
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/CreateMatch'
   *     responses:
   *       201:
   *         description: Match created
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/Match'
   *       400:
   *         description: Validation error
   */
  router.post('/', async (req: Request, res: Response) => {
    const { error, value } = createMatchSchema.validate(req.body)
    if (error) {
      res.status(400).json({ error: error.details[0]?.message })
      return
    }

    const match = await matchService.create(value)
    io.emit('matchCreated', { match })
    res.status(201).json(match)
  })

  /**
   * @swagger
   * /matches/{id}:
   *   patch:
   *     summary: Update match scores
   *     tags: [Matches]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: integer
   *     requestBody:
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/UpdateScore'
   *     responses:
   *       200:
   *         description: Match updated
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/Match'
   *       404:
   *         description: Match not found
   */
  router.patch('/:id', async (req: Request, res: Response) => {
    const { error: idError, value: idValue } = idParamSchema.validate({
      id: parseInt(req.params.id ?? '', 10),
    })
    if (idError) {
      res.status(400).json({ error: idError.details[0]?.message })
      return
    }

    const { error, value } = updateScoreSchema.validate(req.body)
    if (error) {
      res.status(400).json({ error: error.details[0]?.message })
      return
    }

    const match = await matchService.updateScore(idValue.id, value)
    if (!match) {
      res.status(404).json({ error: 'Match not found' })
      return
    }

    io.emit('scoreUpdate', {
      matchId: match.id,
      player1Score: match.player1Score,
      player2Score: match.player2Score,
    })

    res.json(match)
  })

  /**
   * @swagger
   * /matches/{id}/setCurrent:
   *   patch:
   *     summary: Set a match as the current match
   *     tags: [Matches]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: integer
   *     responses:
   *       200:
   *         description: Match set as current
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/Match'
   *       404:
   *         description: Match not found
   */
  router.patch('/:id/setCurrent', async (req: Request, res: Response) => {
    const { error, value } = idParamSchema.validate({ id: parseInt(req.params.id ?? '', 10) })
    if (error) {
      res.status(400).json({ error: error.details[0]?.message })
      return
    }

    const match = await matchService.setCurrent(value.id)
    if (!match) {
      res.status(404).json({ error: 'Match not found' })
      return
    }

    io.emit('currentMatchChanged', { match })

    res.json(match)
  })

  /**
   * @swagger
   * /matches/{id}:
   *   delete:
   *     summary: Delete a match
   *     tags: [Matches]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: integer
   *     responses:
   *       204:
   *         description: Match deleted
   *       404:
   *         description: Match not found
   */
  router.delete('/:id', async (req: Request, res: Response) => {
    const { error, value } = idParamSchema.validate({ id: parseInt(req.params.id ?? '', 10) })
    if (error) {
      res.status(400).json({ error: error.details[0]?.message })
      return
    }

    const deleted = await matchService.delete(value.id)
    if (!deleted) {
      res.status(404).json({ error: 'Match not found' })
      return
    }

    io.emit('matchDeleted', { id: value.id })

    res.status(204).send()
  })

  return router
}

File: packages/server-rust/src/routes/matches.rs
use crate::{
    db::matches as db,
    models::{CreateMatchDto, ErrorResponse, UpdateScoreDto},
    state::AppState,
    websocket::events::{WsEvent, broadcast},
};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use std::sync::Arc;

/// GET /matches
pub async fn list_matches(State(state): State<Arc<AppState>>) -> impl IntoResponse {
    match db::find_all(&state.pool).await {
        Ok(matches) => (StatusCode::OK, Json(serde_json::json!(matches))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// GET /matches/current
pub async fn get_current_match(State(state): State<Arc<AppState>>) -> impl IntoResponse {
    match db::find_current(&state.pool).await {
        Ok(Some(m)) => (StatusCode::OK, Json(m)).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(ErrorResponse::new("No current match")),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// GET /matches/obsData
pub async fn get_obs_data(State(state): State<Arc<AppState>>) -> impl IntoResponse {
    match db::get_obs_data(&state.pool).await {
        Ok(data) => (StatusCode::OK, Json(data)).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// GET /matches/:id
pub async fn get_match(
    State(state): State<Arc<AppState>>,
    Path(id): Path<i64>,
) -> impl IntoResponse {
    match db::find_by_id(&state.pool, id).await {
        Ok(Some(m)) => (StatusCode::OK, Json(m)).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(ErrorResponse::new("Match not found")),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// POST /matches
pub async fn create_match(
    State(state): State<Arc<AppState>>,
    Json(dto): Json<CreateMatchDto>,
) -> impl IntoResponse {
    if dto.player1.trim().is_empty() || dto.player2.trim().is_empty() {
        return (
            StatusCode::BAD_REQUEST,
            Json(ErrorResponse::new("player1 and player2 are required")),
        )
            .into_response();
    }

    match db::create(&state.pool, dto).await {
        Ok(m) => {
            broadcast(&state.ws_clients, WsEvent::MatchCreated { match_data: m.clone() }).await;
            (StatusCode::CREATED, Json(m)).into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// PATCH /matches/:id
pub async fn update_score(
    State(state): State<Arc<AppState>>,
    Path(id): Path<i64>,
    Json(dto): Json<UpdateScoreDto>,
) -> impl IntoResponse {
    match db::update_score(&state.pool, id, dto).await {
        Ok(Some(m)) => {
            broadcast(
                &state.ws_clients,
                WsEvent::ScoreUpdate {
                    match_id: m.id,
                    player1_score: m.player1_score,
                    player2_score: m.player2_score,
                },
            )
            .await;
            (StatusCode::OK, Json(m)).into_response()
        }
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(ErrorResponse::new("Match not found")),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// PATCH /matches/:id/setCurrent
pub async fn set_current(
    State(state): State<Arc<AppState>>,
    Path(id): Path<i64>,
) -> impl IntoResponse {
    match db::set_current(&state.pool, id).await {
        Ok(Some(m)) => {
            broadcast(
                &state.ws_clients,
                WsEvent::CurrentMatchChanged { match_data: Some(m.clone()) },
            )
            .await;
            (StatusCode::OK, Json(m)).into_response()
        }
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(ErrorResponse::new("Match not found")),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}

/// DELETE /matches/:id
pub async fn delete_match(
    State(state): State<Arc<AppState>>,
    Path(id): Path<i64>,
) -> impl IntoResponse {
    match db::delete(&state.pool, id).await {
        Ok(true) => {
            broadcast(&state.ws_clients, WsEvent::MatchDeleted { id }).await;
            StatusCode::NO_CONTENT.into_response()
        }
        Ok(false) => (
            StatusCode::NOT_FOUND,
            Json(ErrorResponse::new("Match not found")),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse::new(e.to_string())),
        )
            .into_response(),
    }
}
